File: server/pipeline/captions.py
from dataclasses import dataclass
from pathlib import Path

import pysubs2

from .transcribe import Word

# Subtitle line-length guideline (broadcast captioning conventions land around
# 32-42 chars/line for a single line at this font size). A long, uninterrupted
# monologue would otherwise become one giant cue with no natural break.
MAX_LINE_CHARS = 42
# Cap how long a single cue sits on screen even if the line is short -- reading
# speed, not just length, bounds a caption's welcome.
MAX_CUE_DURATION = 3.5
# A pause this long between words reads as a new thought; breaking the cue
# there instead of mid-flow matches how captions are actually authored.
BREAK_ON_PAUSE = 0.6

FONT_NAME = "Arial"


@dataclass
class CaptionCue:
	start: float
	end: float
	text: str


def _clean(text: str) -> str:
	# `{` / `}` are ASS override-block delimiters; strip them so stray
	# punctuation in a transcript can't be read as a formatting tag.
	return text.replace("{", "").replace("}", "").strip()


def build_caption_cues(
	words: list[Word],
	max_line_chars: int = MAX_LINE_CHARS,
	max_cue_duration: float = MAX_CUE_DURATION,
	break_on_pause: float = BREAK_ON_PAUSE,
) -> list[CaptionCue]:
	"""Group word-level timestamps into readable caption cues.

	Using word timestamps (rather than the coarser per-turn or per-segment
	timing) is the whole point -- it lets a cue start and end exactly when
	speech does, instead of inheriting a turn's boundaries which can run many
	seconds past the words that justify a caption being on screen.
	"""
	cues: list[CaptionCue] = []
	current: list[Word] = []

	def flush() -> None:
		if not current:
			return
		text = _clean(" ".join(w.text for w in current))
		if text:
			cues.append(CaptionCue(start=current[0].start, end=current[-1].end, text=text))
		current.clear()

	for word in words:
		text = word.text.strip()
		if not text:
			continue

		if current:
			pause = word.start - current[-1].end
			candidate_text = " ".join(w.text for w in [*current, word])
			duration = word.end - current[0].start
			if pause >= break_on_pause or len(candidate_text) > max_line_chars or duration > max_cue_duration:
				flush()

		current.append(word)

	flush()
	return cues


def write_ass(cues: list[CaptionCue], path: Path, frame_w: int, frame_h: int) -> None:
	"""Burn-in caption track sized relative to the export's own frame, so it
	reads the same whether the source is 1080p or 4K."""
	subs = pysubs2.SSAFile()
	subs.info["PlayResX"] = str(frame_w)
	subs.info["PlayResY"] = str(frame_h)

	style = pysubs2.SSAStyle()
	style.fontname = FONT_NAME
	style.fontsize = round(frame_h * 0.045)
	style.bold = True
	style.primarycolor = pysubs2.Color(255, 255, 255)
	style.outlinecolor = pysubs2.Color(0, 0, 0)
	style.borderstyle = 1
	style.outline = max(1, round(frame_h * 0.003))
	style.shadow = 0
	style.alignment = 2  # bottom-centre
	style.marginv = round(frame_h * 0.07)
	subs.styles["Caption"] = style

	for cue in cues:
		subs.events.append(
			pysubs2.SSAEvent(
				start=int(round(cue.start * 1000)),
				end=int(round(cue.end * 1000)),
				text=cue.text,
				style="Caption",
			)
		)

	subs.save(str(path), format_="ass")

File: src/lib/captions.ts
/**
 * Caption cues for the live preview, grouped exactly the way the export
 * groups them.
 *
 * A direct port of `build_caption_cues` in `server/pipeline/captions.py`.
 * There is no shared-code mechanism between the TypeScript and Python
 * runtimes, so the constants below are the same numbers written twice and
 * drift shows up as "the captions I saw aren't the captions I got" -- the
 * same failure mode `personCrop`/`person_crop` and `exportPanes`/
 * `_segment_filter` already guard against, and the reason `captions.test.ts`
 * and `test_captions.py` assert the same cases.
 *
 * `Word` is imported type-only, so this module pulls in nothing at runtime
 * and Node's test runner can load it.
 */
import type { Word } from "@/lib/api";

/** Broadcast captioning lands around 32-42 characters for a single line at
 * this size; without a cap a monologue becomes one endless cue. */
const MAX_LINE_CHARS = 42;
/** Reading speed, not just length, bounds a caption's welcome. */
const MAX_CUE_DURATION = 3.5;
/** A gap this long reads as a new thought, so the cue breaks there rather
 * than mid-flow. */
const BREAK_ON_PAUSE = 0.6;

export interface CaptionCue {
	start: number;
	end: number;
	text: string;
}

/** `{` and `}` delimit ASS override blocks; the export strips them so stray
 * punctuation can't be read as a formatting tag. The preview strips them for
 * the duller reason that it must show the same string. */
function clean(text: string): string {
	return text.replaceAll("{", "").replaceAll("}", "").trim();
}

export function buildCaptionCues(
	words: Word[],
	maxLineChars = MAX_LINE_CHARS,
	maxCueDuration = MAX_CUE_DURATION,
	breakOnPause = BREAK_ON_PAUSE,
): CaptionCue[] {
	const cues: CaptionCue[] = [];
	let current: Word[] = [];

	const flush = () => {
		if (current.length === 0) return;
		const text = clean(current.map((w) => w.text).join(" "));
		if (text) cues.push({ start: current[0].start, end: current[current.length - 1].end, text });
		current = [];
	};

	for (const word of words) {
		if (!word.text.trim()) continue;
		if (current.length > 0) {
			const pause = word.start - current[current.length - 1].end;
			const candidate = [...current, word].map((w) => w.text).join(" ");
			const duration = word.end - current[0].start;
			if (pause >= breakOnPause || candidate.length > maxLineChars || duration > maxCueDuration) flush();
		}
		current.push(word);
	}
	flush();
	return cues;
}

/**
 * The cue on screen at `t`, or null between cues.
 *
 * Binary search rather than a scan: this runs on every `timeupdate`, and a
 * 53-minute episode is ~2,600 cues built from 8,824 words.
 */
export function cueAt(cues: CaptionCue[], t: number): CaptionCue | null {
	let lo = 0;
	let hi = cues.length - 1;
	while (lo <= hi) {
		const mid = (lo + hi) >> 1;
		if (t < cues[mid].start) hi = mid - 1;
		else if (t >= cues[mid].end) lo = mid + 1;
		else return cues[mid];
	}
	return null;
}
